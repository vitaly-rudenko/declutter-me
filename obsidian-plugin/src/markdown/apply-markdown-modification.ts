export function applyMarkdownModification(input: {
    markdown: string,
    type: 'appendLineAfterContent' | 'prependLineBeforeContent'
    section?: string
    skipProperties: boolean
    content: string
}) {
    const { markdown, type, section, skipProperties, content } = input

    const lines = markdown.split('\n')

    let startIndex = 0
    let endIndex = lines.length

    if (section) {
        startIndex = lines.indexOf(section, startIndex)
        if (startIndex === -1) {
            const lastNonEmptyLineIndex = Math.max(0, lines.findLastIndex(line => line.trim() !== ''))
            if (lines[lastNonEmptyLineIndex] !== undefined && lines[lastNonEmptyLineIndex].trim() !== '') {
                lines.splice(lastNonEmptyLineIndex + 1, 0, section)
            } else {
                lines.splice(lastNonEmptyLineIndex, 0, section)
            }

            startIndex = lastNonEmptyLineIndex + 2
        }

        endIndex = lines.findIndex((line, index) => index > startIndex && line.startsWith('#'))
        if (endIndex === -1) {
            endIndex = lines.length
        }
    } else if (skipProperties) {
        if (lines[0] === '---') {
            const propertiesEndIndex = lines.indexOf('---', 1)
            if (propertiesEndIndex !== -1) {
                startIndex = propertiesEndIndex + 1
            }
        }
    }

    if (type === 'appendLineAfterContent') {
        // @ts-ignore
        const lastNonEmptyLineIndex = lines.findLastIndex((line, index) => index >= startIndex && index < endIndex && line.trim() !== '')
        if (lastNonEmptyLineIndex === -1) {
            lines[startIndex] = content
        } else {
            lines.splice(lastNonEmptyLineIndex + 1, 0, content)
        }
    } else if (type === 'prependLineBeforeContent') {
        const firstNonEmptyLineIndex = lines.findIndex((line, index) => index > startIndex && index <= endIndex && line.trim() !== '')
        if (firstNonEmptyLineIndex === -1) {
            if (lines[startIndex] !== undefined && lines[startIndex].trim() !== '') {
                lines.splice(startIndex, 0, content)
            } else {
                lines[startIndex] = content
            }
        } else {
            lines.splice(firstNonEmptyLineIndex, 0, content)
        }
    }

    return lines.join('\n')
}
