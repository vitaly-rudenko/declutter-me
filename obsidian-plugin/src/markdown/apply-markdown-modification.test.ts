import { applyMarkdownModification } from './apply-markdown-modification'

describe('applyMarkdownModification()', () => {
    describe('type: appendLineAfterContent', () => {
        function testBasic(options: Omit<Parameters<typeof applyMarkdownModification>[0], 'markdown' | 'content'>) {
            expect(applyMarkdownModification({
                ...options,
                markdown: '',
                content: 'test'
            })).toBe('test')

            expect(applyMarkdownModification({
                ...options,
                markdown: 'hello world',
                content: 'test',
            })).toBe('hello world\ntest')

            expect(applyMarkdownModification({
                ...options,
                markdown: '\n\n\nhello world\n\n\n',
                content: 'test',
            })).toBe('\n\n\nhello world\ntest\n\n\n')

            expect(applyMarkdownModification({
                ...options,
                markdown: '\n\n\nhello world\n\n\nhello world\nhi there\n\n\n',
                content: 'multiline\ntest',
            })).toBe('\n\n\nhello world\n\n\nhello world\nhi there\nmultiline\ntest\n\n\n')

            expect(applyMarkdownModification({
                ...options,
                markdown: '---\nhello: world\n---',
                content: 'test',
            })).toBe('---\nhello: world\n---\ntest')

            expect(applyMarkdownModification({
                ...options,
                markdown: '---\nhello: world\n---\nhello world',
                content: 'test',
            })).toBe('---\nhello: world\n---\nhello world\ntest')

            expect(applyMarkdownModification({
                ...options,
                markdown: '---\nhello: world\n---\n\n---\n\nhello world\n\n\n',
                content: 'test',
            })).toBe('---\nhello: world\n---\n\n---\n\nhello world\ntest\n\n\n')
        }

        it('defaults', () => {
            testBasic({ type: 'appendLineAfterContent' })
        })

        it('section: "# section"', () => {
            const options = {
                type: 'appendLineAfterContent',
                section: '# section',
                content: 'test',
            } as const

            expect(applyMarkdownModification({
                markdown: '',
                ...options,
            })).toBe('# section\n\ntest')

            expect(applyMarkdownModification({
                markdown: '# section\n\nhello world',
                ...options,
            })).toBe('# section\n\nhello world\ntest')

            expect(applyMarkdownModification({
                markdown: '# section before\n# section\n#section after\n',
                ...options,
            })).toBe('# section before\n# section\ntest\n#section after\n')
        })
    })

    describe('type: prependLineBeforeContent', () => {
        function testBasic(options: Omit<Parameters<typeof applyMarkdownModification>[0], 'markdown' | 'content'>) {
            expect(applyMarkdownModification({
                ...options,
                markdown: '',
                content: 'test'
            })).toBe('test')

            expect(applyMarkdownModification({
                ...options,
                markdown: 'hello world',
                content: 'test',
            })).toBe('test\nhello world')

            expect(applyMarkdownModification({
                ...options,
                markdown: '\n\n\nhello world\n\n\n',
                content: 'test',
            })).toBe('\n\n\ntest\nhello world\n\n\n')

            expect(applyMarkdownModification({
                ...options,
                markdown: '\n\n\nhello world\n\n\nhello world\nhi there\n\n\n',
                content: 'multiline\ntest',
            })).toBe('\n\n\nmultiline\ntest\nhello world\n\n\nhello world\nhi there\n\n\n')

            expect(applyMarkdownModification({
                ...options,
                markdown: '---\nhello: world\n---',
                content: 'test',
            })).toBe('---\nhello: world\n---\ntest')

            expect(applyMarkdownModification({
                ...options,
                markdown: '---\nhello: world\n---\nhello world',
                content: 'test',
            })).toBe('---\nhello: world\n---\ntest\nhello world')

            expect(applyMarkdownModification({
                ...options,
                markdown: '---\nhello: world\n---\n\n---\n\nhello world\n\n\n',
                content: 'test',
            })).toBe('---\nhello: world\n---\n\ntest\n---\n\nhello world\n\n\n')
        }

        it('defaults', () => {
            testBasic({ type: 'prependLineBeforeContent' })
        })

        it('section: "# section"', () => {
            const options = {
                type: 'prependLineBeforeContent',
                section: '# section',
                content: 'test',
            } as const

            expect(applyMarkdownModification({
                ...options,
                markdown: '',
            })).toBe('# section\n\ntest')

            expect(applyMarkdownModification({
                ...options,
                markdown: '# section\n\nhello world',
            })).toBe('# section\n\ntest\nhello world')

            expect(applyMarkdownModification({
                ...options,
                markdown: '# section before\n# section\n#section after\n',
            })).toBe('# section before\n# section\ntest\n#section after\n')
        })
    })

    describe('[regressions]', () => {
        it('should add section correctly', () => {
            expect(applyMarkdownModification({
                type: 'appendLineAfterContent',
                section: '# section',
                markdown: '---\nhello: world\n---\n---\n',
                content: 'test'
            })).toBe('---\nhello: world\n---\n---\n# section\ntest')
        })
    })
})
