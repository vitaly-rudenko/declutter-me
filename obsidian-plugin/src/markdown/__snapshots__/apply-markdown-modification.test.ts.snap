// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations] 1`] = `
"INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: # section before<br># section<br><br><br># section after<br> 1`] = `
"# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: # section before<br># section<br><br>hello world<br>hi there<br># section after<br> 1`] = `
"# section before
# section

hello world
hi there
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: # section before<br># section<br>hello world<br># section after<br> 1`] = `
"# section before
# section
hello world
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: # section<br><br><br> 1`] = `
"# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: # section<br><br>hello world<br>hi there<br> 1`] = `
"# section

hello world
hi there
INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: # section<br>hello world<br> 1`] = `
"# section
hello world
INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br><br> 1`] = `
"---
hello: world
---

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br><br># section before<br># section<br><br><br># section after<br> 1`] = `
"---
hello: world
---


# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br><br># section before<br># section<br><br>hello world<br>hi there<br># section after<br> 1`] = `
"---
hello: world
---


# section before
# section

hello world
hi there
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br><br># section before<br># section<br>hello world<br># section after<br> 1`] = `
"---
hello: world
---


# section before
# section
hello world
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br><br># section<br><br><br> 1`] = `
"---
hello: world
---


# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br><br># section<br><br>hello world<br>hi there<br> 1`] = `
"---
hello: world
---


# section

hello world
hi there
INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br><br># section<br>hello world<br> 1`] = `
"---
hello: world
---


# section
hello world
INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br> 1`] = `
"---
hello: world
---

hello world
hi there
INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section before<br># section<br><br><br># section after<br> 1`] = `
"---
hello: world
---

hello world
hi there
# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section before<br># section<br><br>hello world<br>hi there<br># section after<br> 1`] = `
"---
hello: world
---

hello world
hi there
# section before
# section

hello world
hi there
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section<br><br><br> 1`] = `
"---
hello: world
---

hello world
hi there
# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section<br><br>hello world<br>hi there<br> 1`] = `
"---
hello: world
---

hello world
hi there
# section

hello world
hi there
INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br>hello world<br> 1`] = `
"---
hello: world
---
hello world
INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section before<br># section<br><br><br># section after<br> 1`] = `
"---
hello: world
---
hello world
# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section before<br># section<br>hello world<br># section after<br> 1`] = `
"---
hello: world
---
hello world
# section before
# section
hello world
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section<br><br><br> 1`] = `
"---
hello: world
---
hello world
# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: appendLineAfterContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section<br>hello world<br> 1`] = `
"---
hello: world
---
hello world
# section
hello world
INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations] 1`] = `
"INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: # section before<br># section<br><br><br># section after<br> 1`] = `
"# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: # section before<br># section<br><br>hello world<br>hi there<br># section after<br> 1`] = `
"# section before
# section

hello world
hi there
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: # section before<br># section<br>hello world<br># section after<br> 1`] = `
"# section before
# section
hello world
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: # section<br><br><br> 1`] = `
"# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: # section<br><br>hello world<br>hi there<br> 1`] = `
"# section

INSERTED
hello world
hi there
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: # section<br>hello world<br> 1`] = `
"# section

INSERTED
hello world
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br><br> 1`] = `
"---
hello: world
---

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br><br># section before<br># section<br><br><br># section after<br> 1`] = `
"---
hello: world
---


# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br><br># section before<br># section<br><br>hello world<br>hi there<br># section after<br> 1`] = `
"---
hello: world
---


# section before
# section

hello world
hi there
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br><br># section before<br># section<br>hello world<br># section after<br> 1`] = `
"---
hello: world
---


# section before
# section
hello world
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br><br># section<br><br><br> 1`] = `
"---
hello: world
---


# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br><br># section<br><br>hello world<br>hi there<br> 1`] = `
"---
hello: world
---


# section

INSERTED
hello world
hi there
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br><br># section<br>hello world<br> 1`] = `
"---
hello: world
---


# section

INSERTED
hello world
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br> 1`] = `
"---
hello: world
---

INSERTED
hello world
hi there
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section before<br># section<br><br><br># section after<br> 1`] = `
"---
hello: world
---

hello world
hi there
# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section before<br># section<br><br>hello world<br>hi there<br># section after<br> 1`] = `
"---
hello: world
---

hello world
hi there
# section before
# section

hello world
hi there
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section<br><br><br> 1`] = `
"---
hello: world
---

hello world
hi there
# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br><br>hello world<br>hi there<br># section<br><br>hello world<br>hi there<br> 1`] = `
"---
hello: world
---

hello world
hi there
# section

INSERTED
hello world
hi there
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br>hello world<br> 1`] = `
"---
hello: world
---

INSERTED
hello world
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section before<br># section<br><br><br># section after<br> 1`] = `
"---
hello: world
---
hello world
# section before
# section


# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section before<br># section<br>hello world<br># section after<br> 1`] = `
"---
hello: world
---
hello world
# section before
# section
hello world
# section after

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section<br><br><br> 1`] = `
"---
hello: world
---
hello world
# section

INSERTED
"
`;

exports[`applyMarkdownModification() type: prependLineBeforeContent [permutations]: ---<br>hello: world<br>---<br>hello world<br># section<br>hello world<br> 1`] = `
"---
hello: world
---
hello world
# section

INSERTED
hello world
"
`;
